main.go

```diff
@@ -138,11 +147,18 @@ func setFlags() *managerOpts {
 }
 
 func main() {
-	opts := setFlags()                                // Add our options to flag set.
-	klog.InitFlags(nil)                               // Add klog options to flag set.
-	flag.CommandLine.AddGoFlagSet(goflag.CommandLine) // Merge klog's gofloag flags into the pflags.
+	opts := setFlags() // Add our options to flag set.
+	logsv1.AddFlags(logOptions, flag.CommandLine)
+	flags.AddTLSOptions(flag.CommandLine, &tlsOptions)
+	flag.CommandLine.SetNormalizeFunc(cliflag.WordSepNormalizeFunc)
+	flag.CommandLine.AddGoFlagSet(goflag.CommandLine) // Merge klog's goflag flags into the pflags.
```

Comment:
The main change is that its now using the component-base elements from the Kubernetes project for logs and flags. It adds [flags for configuring webhook TLS settings](https://github.com/kubernetes-sigs/cluster-api/blob/release-1.4/util/flags/tls.go) (`--tls-min-version`, `--tls-cipher-suites`) and [common logging flags](https://github.com/kubernetes/component-base/blob/1faa3ea45285fd34480de3eb54bf7aa5e08413b7/logs/api/v1/options.go#L339) for Kubernetes components

Reply:
Sure:

```
Usage of ./bin/manager:
      --add-dir-header                      If true, adds the file directory to the header of the log messages
      --alsologtostderr                     log to standard error as well as files (no effect when -logtostderr=true)
      --cloud-config-file string            Overrides the default path to the cloud-config file that contains the CloudStack credentials. (default "/config/cloud-config")
      --cloudstackcluster-concurrency int   Maximum concurrent reconciles for CloudStackCluster resources (default 10)
      --cloudstackmachine-concurrency int   Maximum concurrent reconciles for CloudStackMachine resources (default 10)
      --health-probe-bind-address string    The address the probe endpoint binds to. (default ":8081")
      --kubeconfig string                   Paths to a kubeconfig. Only required if out-of-cluster.
      --leader-elect                        Enable leader election for controller manager. Enabling this will ensure there is only one active controller manager.
      --log-backtrace-at traceLocation      when logging hits line file:N, emit a stack trace (default :0)
      --log-dir string                      If non-empty, write log files in this directory (no effect when -logtostderr=true)
      --log-file string                     If non-empty, use this log file (no effect when -logtostderr=true)
      --log-file-max-size uint              Defines the maximum size a log file can grow to (no effect when -logtostderr=true). Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800)
      --log-flush-frequency duration        Maximum number of seconds between log flushes (default 5s)
      --logging-format string               Sets the log format. Permitted formats: "text". (default "text")
      --logtostderr                         log to standard error instead of files (default true)
      --metrics-bind-addr string            The address the metric endpoint binds to. (default "localhost:8080")
      --namespace string                    Namespace that the controller watches to reconcile cluster-api objects. If unspecified, the controller watches for cluster-api objects across all namespaces.
      --one-output                          If true, only write logs to their native severity level (vs also writing to each lower severity level; no effect when -logtostderr=true)
      --skip-headers                        If true, avoid header prefixes in the log messages
      --skip-log-headers                    If true, avoid headers when opening log files (no effect when -logtostderr=true)
      --stderrthreshold severity            logs at or above this threshold go to stderr when writing to files and stderr (no effect when -logtostderr=true or -alsologtostderr=false) (default 2)
      --tls-cipher-suites strings           Comma-separated list of cipher suites for the webhook server. If omitted, the default Go cipher suites will be used. 
                                            Preferred values: TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384, TLS_CHACHA20_POLY1305_SHA256, TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA, TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA, TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305, TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256, TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA, TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA, TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305, TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256, TLS_RSA_WITH_AES_128_CBC_SHA, TLS_RSA_WITH_AES_128_GCM_SHA256, TLS_RSA_WITH_AES_256_CBC_SHA, TLS_RSA_WITH_AES_256_GCM_SHA384. 
                                            Insecure values: TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256, TLS_ECDHE_ECDSA_WITH_RC4_128_SHA, TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA, TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256, TLS_ECDHE_RSA_WITH_RC4_128_SHA, TLS_RSA_WITH_3DES_EDE_CBC_SHA, TLS_RSA_WITH_AES_128_CBC_SHA256, TLS_RSA_WITH_RC4_128_SHA.
      --tls-min-version string              The minimum TLS version in use by the webhook server.
                                            Possible values are VersionTLS10, VersionTLS11, VersionTLS12, VersionTLS13. (default "VersionTLS12")
  -v, --v Level                             number for the log level verbosity
      --vmodule pattern=N,...               comma-separated list of pattern=N settings for file-filtered logging (only works for text log format)
      --watch-filter string                 Label value that the controller watches to reconcile cluster-api objects. Label key is always cluster.x-k8s.io/watch-filter. If unspecified, the controller watches for all cluster-api objects.
      --webhook-cert-dir string             Specify the directory where webhooks will get tls certificates. (default "/tmp/k8s-webhook-server/serving-certs/")
      ```

Reply:
It seems like the changes are more than just adding extra flags. It changes some of the existing flags, which is a breaking change. I'm ok with shipping this with v0.5.0 (not with v0.4.x) but we need to make this evident in the release notes

```diff
 Usage of ./bin/manager:
-      --add-dir-header                      If true, adds the file directory to the header of the log messages
+      --add_dir_header                      If true, adds the file directory to the header of the log messages
       --alsologtostderr                     log to standard error as well as files (no effect when -logtostderr=true)
       --cloud-config-file string            Overrides the default path to the cloud-config file that contains the CloudStack credentials. (default "/config/cloud-config")
       --cloudstackcluster-concurrency int   Maximum concurrent reconciles for CloudStackCluster resources (default 10)
@@ -7,25 +7,18 @@ Usage of ./bin/manager:
       --health-probe-bind-address string    The address the probe endpoint binds to. (default ":8081")
       --kubeconfig string                   Paths to a kubeconfig. Only required if out-of-cluster.
       --leader-elect                        Enable leader election for controller manager. Enabling this will ensure there is only one active controller manager.
-      --log-backtrace-at traceLocation      when logging hits line file:N, emit a stack trace (default :0)
-      --log-dir string                      If non-empty, write log files in this directory (no effect when -logtostderr=true)
-      --log-file string                     If non-empty, use this log file (no effect when -logtostderr=true)
-      --log-file-max-size uint              Defines the maximum size a log file can grow to (no effect when -logtostderr=true). Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800)
-      --log-flush-frequency duration        Maximum number of seconds between log flushes (default 5s)
-      --logging-format string               Sets the log format. Permitted formats: "text". (default "text")
+      --log_backtrace_at traceLocation      when logging hits line file:N, emit a stack trace (default :0)
+      --log_dir string                      If non-empty, write log files in this directory (no effect when -logtostderr=true)
+      --log_file string                     If non-empty, use this log file (no effect when -logtostderr=true)
+      --log_file_max_size uint              Defines the maximum size a log file can grow to (no effect when -logtostderr=true). Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800)
       --logtostderr                         log to standard error instead of files (default true)
       --metrics-bind-addr string            The address the metric endpoint binds to. (default "localhost:8080")
       --namespace string                    Namespace that the controller watches to reconcile cluster-api objects. If unspecified, the controller watches for cluster-api objects across all namespaces.
-      --one-output                          If true, only write logs to their native severity level (vs also writing to each lower severity level; no effect when -logtostderr=true)
-      --skip-headers                        If true, avoid header prefixes in the log messages
-      --skip-log-headers                    If true, avoid headers when opening log files (no effect when -logtostderr=true)
+      --one_output                          If true, only write logs to their native severity level (vs also writing to each lower severity level; no effect when -logtostderr=true)
+      --skip_headers                        If true, avoid header prefixes in the log messages
+      --skip_log_headers                    If true, avoid headers when opening log files (no effect when -logtostderr=true)
       --stderrthreshold severity            logs at or above this threshold go to stderr when writing to files and stderr (no effect when -logtostderr=true or -alsologtostderr=false) (default 2)
-      --tls-cipher-suites strings           Comma-separated list of cipher suites for the webhook server. If omitted, the default Go cipher suites will be used. 
-                                            Preferred values: TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384, TLS_CHACHA20_POLY1305_SHA256, TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA, TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA, TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305, TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256, TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA, TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA, TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305, TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256, TLS_RSA_WITH_AES_128_CBC_SHA, TLS_RSA_WITH_AES_128_GCM_SHA256, TLS_RSA_WITH_AES_256_CBC_SHA, TLS_RSA_WITH_AES_256_GCM_SHA384. 
-                                            Insecure values: TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256, TLS_ECDHE_ECDSA_WITH_RC4_128_SHA, TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA, TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256, TLS_ECDHE_RSA_WITH_RC4_128_SHA, TLS_RSA_WITH_3DES_EDE_CBC_SHA, TLS_RSA_WITH_AES_128_CBC_SHA256, TLS_RSA_WITH_RC4_128_SHA.
-      --tls-min-version string              The minimum TLS version in use by the webhook server.
-                                            Possible values are VersionTLS10, VersionTLS11, VersionTLS12, VersionTLS13. (default "VersionTLS12")
   -v, --v Level                             number for the log level verbosity
-      --vmodule pattern=N,...               comma-separated list of pattern=N settings for file-filtered logging (only works for text log format)
+      --vmodule moduleSpec                  comma-separated list of pattern=N settings for file-filtered logging
       --watch-filter string                 Label value that the controller watches to reconcile cluster-api objects. Label key is always cluster.x-k8s.io/watch-filter. If unspecified, the controller watches for all cluster-api objects.
       --webhook-cert-dir string             Specify the directory where webhooks will get tls certificates. (default "/tmp/k8s-webhook-server/serving-certs/")
```

Reply:
That's fine with me, i believe the CAPI upgrade was not planned to make it in 0.4.x anyway.